[TT-XLA][PyTorch] Training Workload for tiiuae/Falcon3-1B-Base

samiyadev786 opened the pull request:
## Ticket

Closes #212

## Model Training Setting

- **Model:** `Falcon3-1B-Base`
- **Strategy:** LoRA fine-tuning on a single chip (N150)
- **Framework:** PyTorch
- **Task:** Causal language modeling using `wikitext`

## What's Changed

- Added `falcon3_1b` experiment to BOUNTIES folder with training script, config, and README
- Created `WikitextDataset` extending `BaseDataset` in `datasets/torch/BOUNTIES/wikitext/`
- Added `wikitext` to `dataset_utils.py` factory function
- Implemented embedding layer unfreezing via PEFT `modules_to_save` for Falcon3's limited language coverage
- Split `train_step` and `optimizer_step` for optimal XLA graph performance on TT hardware
- Added initial validation logging before training starts
- Configured `max_length=128` for N150 single-chip compatibility

## Experiment Results

### CPU Baseline Results (with Unfrozen Embeddings, 170 steps)

| Metric | Start (Step 0) | End (Step 170) | Trend |
|--------|---------------|----------------|-------|
| train/loss | 2.50 | 2.25 | Decreasing |
| val/loss | 2.50 | 2.32 | Decreasing |
| Trainable params | — | 294M | Includes embed_tokens |

CPU run stopped after 170 steps (~100 minutes) as the full 3-epoch run was estimated to take ~18 hours. The downward trend confirms the fix works.

![CPU train/loss](https://github.com/user-attachments/assets/a825033a-06c7-458e-a3db-6afdc23b71a9)
![CPU val/loss](https://github.com/user-attachments/assets/e169c032-29ae-4f1f-8a58-9dbffc57bc68)

### TT Experiments (N300s, xla:0, pjrt-plugin-tt 0.9.0)

**Unfrozen Embeddings (tt-falcon3-unfrozen-emb-v2):**

| Step | train/loss | val/loss |
|------|-----------|----------|
| 10 | 2.556 | 2.708 |
| 20 | 2.630 | 2.693 |

**Frozen Embeddings (tt-falcon3-frozen-emb-v2):**

| Step | train/loss | val/loss |
|------|-----------|----------|
| 10 | 2.559 | 2.709 |
| 20 | 2.630 | 2.694 |

![TT val/loss and train/loss comparison](https://github.com/user-attachments/assets/bbd2f9d5-6f5b-4b61-9e69-c3b8d19bad55)

### Key Observations

1. **TT and CPU loss values are aligned** — TT starts at ~2.556, CPU at ~2.50 (small difference expected due to numerical precision between CPU and TT XLA device)
2. **Frozen embeddings plateau**: With only 26M trainable LoRA params, the loss oscillates around 2.65-2.71 — the model cannot adapt token representations for languages not in Falcon3's pre-training data (EN, FR, PT, ES)
3. **Unfrozen embeddings improve learning**: With 294M trainable params (LoRA + embed_tokens), the loss shows a steeper, more consistent decline. CPU baseline confirms this trend continues over more steps (2.50 → 2.25 over 170 steps)
4. **No TT errors**: Both experiments ran successfully on TT hardware with pjrt-plugin-tt 0.9.0

## Checklist

- [x] Main training workflow implementation
- [x] CPU baseline testing
- [x] TT device testing (N300s with unfrozen and frozen embeddings)
- [x] W&B loss curves (train/loss + val/loss) for both TT experiments
- [x] Code review fixes (split train_step/optimizer_step, N150 config, logging alignment)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212832436210731

samiyadev786 commented:
@marty1885 Please review this PR.

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3-1b/compare_results.py:
```diff
@@ -0,0 +1,223 @@
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
```
This whole file seems like an overkill to me. A simple plot of CPU and TT loss for experiment could give insight into all these metrics.

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3-1b/README.md:
```diff
@@ -0,0 +1,334 @@
+# Falcon3-1B-Base LoRA Training on TT-N150
```
Thanks for making this `README.md` really descriptive.

I would keep this file in a similar format as for other experiments. So only related to the specific experiment data and model.

We can split the other part into a `README.md` in the top level `bounty` folder, so that other users that are doing bounties can have a notion of what is expected to complete the whole experiment properly.

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3-1b/wikitext_dataset.py:
```diff
@@ -0,0 +1,212 @@
+from configs import Falcon3TrainingConfig
+
+
+class WikitextDataset(Dataset):
```
We need to extend the main structure we have in the repo. (`BaseDataset`)

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3-1b/utils.py:
```diff
@@ -0,0 +1,674 @@
+@cpu_fallback(
+    operation_name="custom_cross_entropy",
+    reason="Custom cross-entropy may have compilation issues on TT",
+    issue_url="https://github.com/tenstorrent/tt-xla/issues/1993",
```
Issue you linked is realted to multi-chip problems. You are running a single chip experiment.

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3-1b/train_falcon3_lora.py:
```diff
@@ -0,0 +1,631 @@
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+# Conditional import for TT-XLA
+try:
```
Imports should be stable, we should have them in our environments.

mmilosevicTT reviewed (commented):
Hey @samiyadev786. Thanks for taking the initiative to work on this BOUNTY.

I have left comments on your PR. The general structure is going into the direction we want to have. Comments should guide you into a direction we expect. To make sure you can focus addressing of the comments provided, I am providing a short summary of the main things that need to be changed:
- We need to refactor `wikitext` dataset to datasets folder. Please take a look at [dataset](https://github.com/tenstorrent/tt-blacksmith/blob/main/blacksmith/datasets/torch/sst2/sst2_dataset.py)
- We need to refactor the experiment to follow the main structure. There are a lot of repetitions of the code already present in other experiments. Please take a look at [model](https://github.com/tenstorrent/tt-blacksmith/blob/main/blacksmith/experiments/torch/gemma/test_gemma_finetuning.py) for more info on how the experiment should be structured.
- We need to refactor `README.md` and to have one specific file like [gemma experiment](https://github.com/tenstorrent/tt-blacksmith/blob/main/blacksmith/experiments/torch/gemma/README.md). Apart from that we can add a top level file for how BOUNTIES should be done/run so that other community contributors can contribute in less time.
- We need to address the problems with loss, having a final loss of 2.15 is quite huge for a long fine-tuning experiment. This suggest some loss calculation is probably not done in a proper way.
- We need to refactor the code for generating report into utils as well. I am not a fan of keeping results in the repo, but having scripts that make it easier to generate reports/plots could be helpful for other contributors.
- We need to attach results to the PR description as well, when we remove them from the repo.

In general the work you have done is amazing, but there are some polishing things we need to do in order to keep the repo in a consistent state.

Please rename the PR title to what the issue title is https://github.com/tenstorrent/tt-blacksmith/issues/212. Also include `Closes (issue link)` to your PR description so that the PR is automatically linked to the proper issue. Could you please rework the PR description to follow the main template we have for the PR structure. (You can take a look at how these suggestions should be done in here https://github.com/tenstorrent/tt-blacksmith/pull/303).

mmilosevicTT commented on blacksmith/datasets/torch/BOUNTIES/wikitext/wikitext_dataset.py:
```diff
@@ -0,0 +1,161 @@
+            return_attention_mask=True,
+        )
+
+    def _group_texts(self, examples: Dict[str, List]) -> Dict[str, List]:
```
Great idea.

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3-1b/test_falcon3_finetuning.py:
```diff
@@ -0,0 +1,214 @@
+This script implements LoRA fine-tuning of tiiuae/Falcon3-1B-Base
+on the Wikitext-2 dataset using TT-XLA.
+
+Usage:
```
No need for this comment, we already have it in `README.md`.

mmilosevicTT reviewed (commented):
@samiyadev786 Great work!

The structure of the code is great and I left only few finishing comments on that.

Could you please make sure to include a plot of the loss you get when you run CPU/TT experiment? This is something we need to confirm the training is actually working as expected. For example, look at the PR description in here https://github.com/tenstorrent/tt-blacksmith/pull/303.

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3-1b/test_falcon3_finetuning.yaml:
```diff
@@ -0,0 +1,60 @@
+# Dataset settings
+dataset_id: "wikitext"
```
We need to have a separate `config.py` file along side the `.yaml` one. Please take a look at other experiments.

samiyadev786 commented:
> Hi @samiyadev786, just checking if you've tried unfreezing the embedding layer and checked the CPU/TT loss curves. Any updates?

@minaliuTT I am on it.

samiyadev786 commented:
Hi @mmilosevicTT,

Thanks for the suggestion about unfreezing the embedding layer. I've implemented it using PEFT's `modules_to_save=["embed_tokens"]` in the LoRA config, which allows the embedding layer to be fully trained alongside the LoRA adapters.

## Changes Made

- Added `unfreeze_embeddings` config option to `TrainingConfig` (default `True` for Falcon3, `False` for other experiments)
- Modified `_apply_lora()` in `hf_models.py` to pass `modules_to_save` when `unfreeze_embeddings` is enabled
- Updated YAML config and README documenting the rationale (Falcon3 limited to EN/FR/PT/ES, wikitext contains other language tokens)
- Fixed `WikitextDataset` to implement `_get_dataloader()` matching `BaseDataset` abstract method

## CPU Results (with unfrozen embeddings)

| Metric | Start | Current (Step 150) | Trend |
|--------|-------|---------------------|-------|
| train/loss | 2.55 | 2.25 | Decreasing |
| val/loss | 2.52 | 2.30 | Decreasing |
| Trainable params | — | 294M | Includes embed_tokens |

Loss is now showing proper learning with a clear downward trend, unlike before where it plateaued at ~2.4 with frozen embeddings.

I stopped the CPU run after 150 steps (~100 minutes) as the full 3-epoch run was estimated to take ~18 hours on CPU. The downward trend is clear enough to confirm the fix is working. Will run the full experiment on TT hardware next for faster results and attach both CPU and TT loss plots.

<img width="1920" height="906" alt="image" src="https://github.com/user-attachments/assets/a825033a-06c7-458e-a3db-6afdc23b71a9" />

<img width="1920" height="913" alt="image" src="https://github.com/user-attachments/assets/e169c032-29ae-4f1f-8a58-9dbffc57bc68" />

## Next Step:

- Running TT experiment next — will attach loss plots once complete.

samiyadev786 commented:
Hi @mmilosevicTT,

I've completed the TT experiments with and without embedding layer unfreezing as requested. Here are the results:

## TT Experiment Results (Both runs on N300s with pjrt-plugin-tt 0.9.0)

### Comparison: Frozen vs Unfrozen Embeddings on TT

Both TT runs overlaid show that the loss curves are very similar in the early steps (0-40), starting at 3.20 and dropping to ~2.65. The frozen embeddings run (pink) shows oscillation and plateaus after step 10 (2.46 → 2.65 → 2.71 → 2.67), while the unfrozen embeddings run (green) maintains a steadier downward trend.

<img width="1821" height="674" alt="image" src="https://github.com/user-attachments/assets/6df6a378-26b0-4ef6-819f-3a8bbf80a577" />

### TT with Frozen Embeddings (`unfreeze_embeddings: False`)
| Step | train/loss |
|------|-----------|
| 0 | 3.20 |
| 10 | 2.46 |
| 20 | 2.66 |
| 30 | 2.71 |
| 40 | 2.67 |

- val/loss: 2.72
- Trainable params: **26M** (LoRA only)
- Loss drops initially but **oscillates and plateaus around 2.65-2.71** after step 10

<img width="706" height="723" alt="image" src="https://github.com/user-attachments/assets/edcff335-4200-4645-8116-123d287465b6" />

### TT with Unfrozen Embeddings (`unfreeze_embeddings: True`)
| Step | train/loss |
|------|-----------|
| 0 | 3.20 |
| 10 | 2.46 |

- val/loss: 2.72
- Trainable params: **294M** (LoRA + embed_tokens)
- Loss shows a **steeper, more consistent downward trend**

<img width="751" height="726" alt="image" src="https://github.com/user-attachments/assets/8153273a-ef31-4b23-8363-afb8877ccba2" />

### CPU Baseline (with Unfrozen Embeddings, 170 steps)
- train/loss: 2.50 → 2.25 (clear downward trend)
- val/loss: 2.50 → 2.32
- Confirms unfrozen embeddings enable proper learning

## Key Observations

1. **Frozen embeddings plateau**: With only 26M trainable LoRA params, the loss oscillates around 2.65-2.71 after the initial drop — the model cannot adapt token representations for languages not in Falcon3's pre-training data (EN, FR, PT, ES).
2. **Unfrozen embeddings improve learning**: With 294M trainable params (LoRA + embed_tokens), the loss shows a steeper, more consistent decline. The CPU baseline confirms this trend continues over more steps (2.50 → 2.25 over 170 steps).
3. **No TT errors**: Both experiments ran successfully on TT hardware with pjrt-plugin-tt 0.9.0.

## Implementation

- `modules_to_save=["embed_tokens"]` in PEFT LoRA config unfreezes the embedding layer
- TT experiments used `batch_size=1`, `max_length=256` to fit within N300s DRAM
- Default config (`batch_size=4`, `max_length=512`) causes OOM on N300s

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3_1b/test_falcon3_finetuning.py:
```diff
@@ -0,0 +1,198 @@
+
+    with torch.no_grad():
+        for batch in tqdm(val_data_loader, desc="Validation"):
+            batch = device_manager.prepare_batch(batch)
```
Could you please rework validation like in llama experiment on xla? We had memory issues when we used prepare_batch in here, as TT device was holding more tensors than needed.

mmilosevicTT commented on blacksmith/datasets/torch/BOUNTIES/wikitext/wikitext_dataset.py:
```diff
@@ -0,0 +1,146 @@
+        # Filter out empty examples
+        raw_dataset = raw_dataset.filter(lambda example: len(example["text"].strip()) > 0)
+
+        # Tokenize and chunk the dataset
```
```suggestion
        # Tokenize and chunk the dataset.
```

mmilosevicTT commented on blacksmith/datasets/torch/BOUNTIES/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
```
```suggestion
# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
```

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3_1b/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
```
```suggestion
# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
```

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3_1b/test_falcon3_finetuning.yaml:
```diff
@@ -0,0 +1,64 @@
+
+# Model settings
+model_name: "tiiuae/Falcon3-1B-Base"
+max_length: 512
```
ditto as in `config.py`

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3_1b/test_falcon3_finetuning.py:
```diff
@@ -0,0 +1,197 @@
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
```
```suggestion
# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
```

mmilosevicTT reviewed (commented):
Hey @samiyadev786,

Amazing work on this issue 😄.

I only have three last things that you need to do for the PR to align with how we do experiments. To help you out I will summarise them here:
- Performance problem I mentioned in `train_step` where you need to split `train_step` and `optimizer_step`.
- Problem with experiment code alignment that is happening in the main code.
- The config setup that can actually run on N150, so we properly represent what can be run on that chip.

All other comments are nits and are just fine polishing. Please let me know when you manage to align these last few things so we can complete this issue.

mmilosevicTT commented on blacksmith/experiments/torch/BOUNTIES/falcon3_1b/test_falcon3_finetuning.py:
```diff
@@ -0,0 +1,260 @@
+                    model.train()
+                    logger.log_metrics(
+                        {"epoch": epoch + 1, "val/loss": avg_val_loss}, step=global_step
+                    )
```
```suggestion
                if global_step % config.steps_freq == 0:
                    avg_loss = (
                        running_loss / config.steps_freq
                        if global_step > 0
                        else running_loss
                    )
                    logger.log_metrics(
                        {"train/loss": avg_loss},
                        commit=False,
                        step=global_step,
                    )
                    running_loss = 0.0

                if global_step % config.val_steps_freq == 0:
                    model.eval()
                        val_loss = validate(
                            model,
                            eval_dataloader,
                            loss_fn,
                            device_manager.device,
                            config,
                            logger,
                            tokenizer,
                        )
                        logger.log_metrics({"val/loss": val_loss}, commit=False, step=global_step)
                        model.train()

                logger.log_metrics({}, commit=True, step=global_step)
```

mmilosevicTT reviewed (commented):
Great work, left few polishing comments and I will be ok with merging. Also, I enabled workflows on you PR so please make sure they all pass.

Pinging @vladimirjovanovicTT, if you want to add any additional comments before merging.

samiyadev786 commented:
> LGTM @samiyadev786
> 
> Thanks for the amazing work.

@mmilosevicTT My Pleasure
